### DISCUSSION TITLE
Zitadel API

### DISCUSSION THREAD

**Discussion by MLStoltzenburg** [zitadel/zitadel#4036] [Q&A] [answered]
We are unsure how to access the Zitadel API on our own domain.
We followed the documentation but were not successful.
We get the "Invalid Audience" error when accessing a specific domain, but we were successful when we tried from the Zitadel Instance's main domain.

What are we doing wrong?

Is there a complete example of how we can do this using a secondary domain configured on the same instance?

The link of APIs: https://editor.swagger.io/?url=https://auth.stolzlabs.com/openapi/v2/swagger/auth.swagger.json.

Example:

 scopes = [
    'openid',
    'profile',
    'email',
    'offline_access',
    'urn:zitadel:iam:org:domain:primary:stolzlabs.com',
    'urn:zitadel:iam:org:project:id:XXXXXXXXXXXX:aud',
  ];

curl -X 'GET' \
  'https://auth.stolzlabs.com/auth/v1/users/me' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer [access token Bearer Token type]'

{"code":16, "message":"invalid audience (APP-Zxfako)", "details":[{"@type":"type.googleapis.com/zitadel.v1.ErrorDetail", "id":"APP-Zxfako", "message":"invalid audience"}]

Thank you for your help!

**Comment by fforootd** [comment]
Hi, good question

Most of the times you primary want to access the [management API](https://docs.zitadel.com/docs/apis/introduction#management). the auth API only exposes the currently logged in users functions.
In which case you want your client to request the scope `urn:zitadel:iam:org:project🆔XXXXXXXXXXXX:aud` where you replace the XXXX with the project ID of ZITADEL itself.

E.g in my test instance this one:

![image](https://user-images.githubusercontent.com/9879976/180603630-17b16a08-8a29-4741-9b12-80d3a18e3cdc.png)

Also you want to add the `x-zitadel-orgid` header for ZITADEL if you want to specify a specific organization (the context)

What language are you using, maybe we can improve the quickstarts to get you started more easily.

CC @hifabienne

**Comment by MLStoltzenburg** [comment]
Hey,

I'm trying to access this API via DART/Flutter as I would like users to be able to change their information through the app!

I did another test with your information:

Scopes:
scopes = [
    'openid',
    'profile',
    'email',
    'offline_access',
    'urn:zitadel:iam:org:domain:primary:stolzlabs.com',
    'urn:zitadel:iam:org:project:id:168582863900508873:aud',
  ];

Command:
curl -X 'GET' \
  'https://auth.stolzlabs.com/auth/v1/users/me' \
  -H 'accept: application/json' \
  -H 'x-zitadel-orgid: 168599856586162905' \
  -H 'Authorization: Bearer [token]'
 
Same result:
{"code":16, "message":"invalid audience (APP-Zxfako)", "details":[{"@type":"type.googleapis.com/zitadel.v1.ErrorDetail", "id":"APP-Zxfako", "message":"invalid audience"}]}

Screenshot:
<img width="1235" alt="Captura de Tela 2022-07-23 às 11 41 05" src="https://user-images.githubusercontent.com/2595026/180609888-88a25600-0d4b-4e6d-8a32-b136c4c10d00.png">


<img width="1258" alt="Captura de Tela 2022-07-23 às 11 38 13" src="https://user-images.githubusercontent.com/2595026/180609793-a1cf4ce5-64a7-4eab-9aa1-31906ac5b15d.png">

**Reply by fforootd** [reply]
Hm strange

I need to check this with @livio-a  on Monday

**Reply by MLStoltzenburg** [reply]
Ok!

If you need more information, I can send to you! :-)
We are trying to switch from Keycloak to Zitadel in our solution!

**Reply by fforootd** [reply]
Love to hear this. I will definitely come back to you!

**Reply by fforootd** [reply]
Hi again

I have two questions to better understand the problem here.

1) By accessing a specific domain you mean a domain that's attached to an organization and invoked through the primary domain scope, right?

> We get the "Invalid Audience" error when accessing a specific domain, but we were successful when we tried from the Zitadel Instance's main domain.

So if you leave the primary domain from the scope, it does work?

2) The login with you client (you app in this case) works fine and only while calling the Auth API with the recieved access_token you get the invalid audience error, right?

**Reply by MLStoltzenburg** [reply]
Hi,

Answer 1:

I have an instance where I have two Organizations Zitadel and Stolz Labs.
My users are at Stolz Labs Organization, but I would like my user to access the Auth-Api, but I understood that API is in the Zitadel Organization, so I put AUD from Zitadel and org id from Stolz Labs.

My idea is the user himself gets his information as "/users/me" or change his password.

I don't know if I did this correctly, but I understand that APIs are at Zitadel Organization.


Answer 2:

You are right! It is the behavior that I have.

**Reply by fforootd** [reply]
@livio-a @hifabienne can it be that the two scopes somehow end up influencing each other.

It somewhat sounds to me that way.

**Reply by hifabienne** [reply]
We just tested this by our self. And it worked for us. At the moment we can't see what is different to you.

This is how our setup looked like:
Organisation 1: Name: ACME, Domain: acme-adsif.zitadel.cloud
- Projects: ZITADEL 

Organisation 2: Name: Test, Domain: test.acme-gzoe4x.zitadel.cloud
- Projects: Test, App: test
- User: test@test.acme-gzoe4x.zitadel.cloud

We logedin with the test user of organisation 2 as client id we set the id of the app "test" in organisation 2.

our scopes: openid urn:zitadel:iam:org:project:id:165617850743128321:aud urn:zitadel:iam:org:domain:primary:test.acme-gzoe4x.zitadel.cloud

The id in the first scope is the id of the ZITADEL project in organisation 1 and the primary domain is from organisation 2.

Can you have a look at your database after you did a request? 
`select * from auth.tokens order by creation_date desc limit 5;`
With this you should see all the request, in the "scopes" column you should see that the audience scope was requested and in the "audience" column the requested id should be included.

Because the error you get, is shown when this is not included in the request.

**Reply by hifabienne** [reply]
You could also try to make an introspect call and send us what is included in the response. There you should see the same as in the database

**Comment by MLStoltzenburg** [comment, marked as answer]
I don't know why, but now it works!
I did not change anything.

All good! At least now it's working! :-)

select:
```
          id         |         creation_date         |          change_date          |   resource_owner   |        application_id        |   user_agent_id    |      user_id       |          expiration           | sequence |                                                                    scopes                                                                    |                                                             audience                                                             | preferred_language |  refresh_token_id  | is_pat |    instance_id
---------------------+-------------------------------+-------------------------------+--------------------+------------------------------+--------------------+--------------------+-------------------------------+----------+----------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------+--------------------+--------------------+--------+---------------------
  172012649326837963 | 2022-07-28 15:52:08.604048+00 | 2022-07-28 15:52:08.604048+00 | 168599856586162905 | 168675490624766681@stolzlabs | 171172740135387320 | 169696726758720218 | 2022-07-29 03:52:08.603667+00 |     1770 | {openid,profile,email,offline_access,urn:zitadel:iam:org:domain:primary:stolzlabs.com,urn:zitadel:iam:org:project:id:168582863900508873:aud} | {168965004823953591@stolzlabs,168675490624766681@stolzlabs,168675304296940249,168582863900508873}                                | en                 | 172012649310060747 | false  | 168582863850046153


```
scope:
```
  scope = [
    'openid',
    'profile',
    'email',
    'offline_access',
    'urn:zitadel:iam:org:domain:primary:stolzlabs.com',
    'urn:zitadel:iam:org:project:id:168582863900508873:aud',
  ];
```

I got the access token and ran:
```
curl -X 'GET' \
  'https://auth.stolzlabs.com/auth/v1/users/me' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer 2fWXad2eiIjCrNzVvpIOjYGJqpij1M8a8U-zvCDWuZuvYtC7FuFpE5VNCvUyyhn0cCLdCQg'
```
Result:
```
{
   "user":{
      "id":"169696726758720218",
      "details":{
         "sequence":"1623",
         "creationDate":"2022-07-12T16:25:32.469416Z",
         "changeDate":"2022-07-26T16:23:18.861038Z",
         "resourceOwner":"168599856586162905"
      },
      "state":"USER_STATE_ACTIVE",
      "userName":"XXXXXX",
      "loginNames":[
         "XXXXXXXX"
      ],
      "preferredLoginName":"XXXXXX",
      "human":{
         "profile":{
            "firstName":"Marcelo",
            "lastName":"Lemos",
            "displayName":"Marcelo Lemos",
            "preferredLanguage":"en",
            "gender":"GENDER_MALE"
         },
         "email":{
            "email":"XXXXXXXXXXXX",
            "isEmailVerified":true
         },
         "phone":{
            
         }
      }
   }
}
```

Thanks for help!
